### DISCUSSION TITLE
using cross account eventbridge

### DISCUSSION THREAD

**Discussion by SteveL1** [cloud-custodian/cloud-custodian#7448] [General]
### Ask your question

I am currently deploying dozens of policies to dozens of accounts.  The accounts are managed and paid for by individual teams but I responsible for ensuring that all the accounts follow company guidelines and this is causing a couple of concerns.

Firstly the costs (low as they are) are born by the account holders and not by my team.  We have had a case where a policy was incorrect and was being launched recursively, meaning that it had millions of invocations when only one was needed!  It was my fault, but the team affected was billed.  We have tagged the functions with a cost centre to cross-charge the fees but it creates an unnecessary overhead on someone's time to process this.

Secondly, and more importantly, as we don't manage the accounts there is a possibility that someone can remove or update the lambda functions.  I redeploy the functions daily but it someone could easily schedule the removal of the functions.

Also, I have found that mugc.py isn't perfect and will sometimes leave old policies behind and reviewing all of the accounts when there is a policy name change can be quite time consuming.

Is there a way to either send eventbridge messages to a central account where the lambda's exist or for custodian to assume a role in a different account and they can be processed there while still remediating in the source account?

**Comment by kapilt** [comment]
yes, you can use member-role in lambda execution modes to cross account role assume. member-role is interpolated with the account id as a template. the setup of the event bridge forwarding is done out of band as a one time activity.

**Comment by SteveL1** [comment]
Hi,

Thanks so much for replying to me.

I wonder if you could expand on how to implement this, i'm not sure what you mean by "member-role is interpolated with the account id as a template"

Setting up eventbridge shouldn't be an issue, but I just need to get Custodian to remediate in the correct account

Thanks

**Comment by ajkerrigan** [comment]
Hi @SteveL1, the idea is that your policy's `mode` block can define both `role` and `member-role`:

```yaml
    mode:
        type: cloudtrail
        events:
          - RunInstances
        role: arn:aws:iam::{account_id}:role/cloud-custodian
        member-role: arn:aws:iam::{account_id}:role/cloud-custodian-remediations
```

Custodian will provision a Lambda function using `role`, filling in `{account_id}` based on the credentials you use with `custodian run`.

When that policy Lambda fires it will assume `member-role`, filling in `{account_id}` based on the account ID in the triggering event.

This does require some out of band setup on your end:
* The role you specify in `role` must be able to assume your `member-role` in all target accounts.
* You need to set up EventBridge event forwarding from your target accounts to the account where your policy Lambda functions live.

I hope that helps, but let us know if you have other questions. Thanks!

**Reply by msi-arnold** [reply]
hi @ajkerrigan 

since this is cross account using eventbridge, and in cloud custodian is using type: cloudtrail does it mean that this will also check eventbridge for cloudtrail event from other account?

thanks

**Reply by ajkerrigan** [reply]
Yes that's the idea, as long as you've set up eventbridge forwarding rules to send those cloudtrail events to your central account. Which happens outside of Custodian's awareness/control.

**Reply by msi-arnold** [reply]
hi @ajkerrigan thanks for the response,

however, if it setup like this:
role: arn:aws:iam::{account_id}:role/cloud-custodian
member-role: arn:aws:iam::{account_id}:role/cloud-custodian-remediations

how cloud custodian now what account it was executed and what account is a member account? because I would assume that the {account_id} would be the same account number from account.yml
